### DISCUSSION TITLE
Connecting a click event on a widget to a function.

### DISCUSSION THREAD

**Discussion by JoeZiminski** [Textualize/textual#3680] [Q&A] [answered]
Is it possible to connect a function to a click event on a specific widget? In particular, I am interested in linking a function execution to double-clicking on a particular `Input` widget. As far as I can see, the `_on_click()` method of `Input` does not have any kind of [emit signal](https://github.com/Textualize/textual/discussions/186) and the generic `on_click` method does not need to hold information about the clicked widget. I presume it is possible based on the x,y position of the click and widget but was wondering if there is an easier way. 

Apologies if this is an obvious question,  I could not find an answer after a reasonably thorough search of docs / issues but I may be looking for the wrong thing!

**Comment by davep** [comment, marked as answer]
If it's a specific widget you want to do this for, you could subclass that widget and handle it from there, adding your own `Clicked` message. For example:

```python
from __future__ import annotations
from dataclasses import dataclass

from textual import on
from textual.app import App, ComposeResult
from textual.events import Click
from textual.message import Message
from textual.widgets import Input, Log

class ClickableInput(Input):

    @dataclass
    class Clicked(Message):
        input: ClickableInput

    def _on_click(self, _: Click) -> None:
        self.post_message(self.Clicked(self))


class InputClickApp(App[None]):

    def compose(self) -> ComposeResult:
        yield ClickableInput()
        yield Log()

    @on(ClickableInput.Clicked)
    def log_click(self, event: ClickableInput.Clicked) -> None:
        self.query_one(Log).write_line(f"{event!r}")

if __name__ == "__main__":
    InputClickApp().run()
```

More generally, you *could* use `get_widget_at` to get the widget that was clicked, and use it at a higher level in the DOM; unfortunately though if the widget being clicked is processing and stopping the event, it'll never bubble up. For example:

```python
from textual.app import App, ComposeResult
from textual.events import Click
from textual.widgets import Input, Label, Button

class ClickAnywhereApp(App[None]):

    def compose(self) -> ComposeResult:
        for n in range(10):
            yield Input(id=f"input-{n}")
            yield Label(f"Label {n}", id=f"label-{n}")
            yield Button(f"Button {n}", id=f"button-{n}")

    async def _on_click(self, event: Click) -> None:
        self.notify(f"{self.get_widget_at(event.screen_x, event.screen_y)[0]}")

if __name__ == "__main__":
    ClickAnywhereApp().run()
```

You'll always get a notification from the screen and the labels, but never from the `Button` or the `Input` (actually, in testing, I was seeing the occasional one from the `Input` which is kind of odd).

But, if as you say:

> I am interested in linking a function execution to double-clicking on a particular Input widget

then you could roll your own subclass of `Input` like the above and you should be good to go. Note though that Textual doesn't have a double-click event.

**Comment by JoeZiminski** [comment]
Thanks a lot @davep for the detailed and very prompt response! I will try this out today and let you know if anything comes up. Cheers!

**Comment by JoeZiminski** [comment]
Thanks @davep works like a dream! I had one question, I see it is not necessary to also call the method on the super class e.g.
`super(ClickableInput, self)._on_click()` to get the superclass functionality. It this becuse it is an `async` function?

**Reply by davep** [reply]
Not because of `async`, but because [Textual handles that on purpose for you](https://textual.textualize.io/guide/events/#default-behaviors). Which may lead to the next obvious question: [how to stop it doing that when you don't want it to](https://textual.textualize.io/guide/events/#preventing-default-behaviors).

**Reply by JoeZiminski** [reply]
That makes sense, that is a nice design! thanks for all your help.
